_re2.cc

```diff
@@ -567,7 +600,16 @@ _group_span(MatchObject2* self, long idx, Py_ssize_t* o_start, Py_ssize_t* o_end
     *o_end = -1;
     return false;
   }
-  Py_ssize_t start = piece.data() - PyString_AS_STRING(self->string);
+  Py_ssize_t start;
+#if PY_MAJOR_VERSION >= 3
+  if (PyBytes_Check(self->string)) {
```

Comment:
Yeah that's worth some tests to explain what I was looking for. Basically, patterns should only be passed as "strings" (`unicode` in py3, `str` in py2), which matches re2's requirements that the data be encoded as either utf-8 or latin1 (we've left it at the default of utf8). Data can come in as bytes or unicode so we can use a binary-capable regex to search a binary blob of data. I haven't checked (needs test!) whether or not matches come back as the same type as the data, but I think they do.

Reply:
This seems reasonable to me.  We could either explicitly remember that we were called on a string/bytes, or just (as this version does) check every time since we're already holding a reference to the target.  If it turns out to hurt performance, it's very easy to introduce the boolean later.